### PR TITLE
ACE 100 - Disable authentication

### PR DESCRIPTION
**Pull request by hessam61**
Verifications steps:
```
$ cd docker-examples/development-docker-compose
$ docker-compose up
```
This PR is intended for internal review and adds a `noauth` user object with admin access to disable authentication. It also removes all user-specific items from React application. The addition of new docker compose is intended for system testing, and will probably not be a part of contributing PR. 

NOTE: This branch is in-sync with upstream/master since the original author added a new user menu and I wanted to include that in this system test. However, our forked master branch is outdated, but we should be able to merge once that branch is synced up with upstream/master. This explains the extra commits. If this is too hard to review, please let me know to sync up our master first, then create another PR.

TODO: This is not required for submitting the PR, but I would like to work on a different version of docker compose that utilizes `npm run start` for the react app in order to enhance local development. This might be useful if we continue to contribute more to this project.

### CODE REVIEW

**Comment by bruth** [comment]
@hessam61 Can you or @murphyke rebase `master` to match the origin repo? That way this PR only shows commits relevant to your changes.

**Comment by hessam61** [comment]
@bruth Sure that makes sense actually. I can rebase master with upstream, and if needed recreate the PR to remove unnecessary history.

**Comment by hessam61** [comment]
@murphyke @bruth I rebased our master with upstream. Things should be cleaner to review.

File: client/src/app-header/AppHeaderUser.js
Comment (bruth) [inline client/src/app-header/AppHeaderUser.js]: Style nit, add a space after `if`.

File: client/src/app-header/AppHeaderUser.js
Comment (bruth) [inline client/src/app-header/AppHeaderUser.js]: This a readability preference in general, but since the `if` block above returns and this `else` block is not conditioned, the `else` can be removed and simply return the `<div ...`.

So like this:

```jsx
if (currentUser._id === "noauth") {
  return null;
}

return <div className={styles.style}> {currentUser.name || currentUser.email}</div>;
```

File: server/middleware/disable-auth.js
Comment (bruth) [inline server/middleware/disable-auth.js]: This seems to be indented one too many.

**Review by bruth — CHANGES_REQUESTED** [review]
Thanks for rebasing against upstream master, however it still includes Kevin's merged BigQuery commit on this repo's master. Not a big deal for the review (easy to ignore those files), but this should be rebased before merging.

In terms of the code itself, I left a few minor code nits, but looks good overall! I am going to trying running it locally shortly.

**Comment by bruth** [comment]
Screen Shot 2020-03-11 at 4 29 33 PM

I am observing this error. The SQL still appears to execute (at least in the case of creating the `test` table in the SQLite database I tried). But it seems to be complaining about `userEmail` somewhere. Maybe it is set as a null value and should be a empty string to bypass the issue?

Something else I noticed which may not be an issue is that I was still able to go to the `/signin` page. If its straightforward, maybe this should be auto-redirected to the main page when auth is disabled.

In terms of running the server, I ended up following Kevin's directions from his PR:

- Run `scripts/build.sh` (this is per DEVELOPER-GUIDE.md).
- `export DISABLE_AUTH=true`
- `cd server && node server.js --dbPath ../db --port 3010`
- Surf to http://localhost:3010

When I tried the docker-compose approach, it was hanging on the one `RUN` step.

File: client/src/app-header/AppHeaderUser.js
Comment (murphyke) [reply client/src/app-header/AppHeaderUser.js]: Regarding all superficial style issues like whitespace, positioning of braces, etc, SQLPad uses `prettier`. In the top level, there are `fixlint` and `lint` scripts, so you can do `npm run lint` and `npm run fixlint` -- the latter will rewrite the files. 

VS Code and most editors have a Prettier plugin/integration.

Comment (hessam61) [reply client/src/app-header/AppHeaderUser.js]: Good to know! I applied the linter for all files.

File: client/src/app-header/AppHeaderUser.js
Comment (hessam61) [reply client/src/app-header/AppHeaderUser.js]: ✅

File: server/middleware/disable-auth.js
Comment (hessam61) [reply server/middleware/disable-auth.js]: Linter should have fixed this.

**Comment by hessam61** [comment]
@bruth That is interesting, docker-compose takes a long time for me too. This wasn't the case few hours ago, it runs but takes a really long time now. I wonder if one of the NPM repositories has a slow connection.

As for the error message, I did run into that before but wasn't able to replicate that. I think one of the node dependencies is actually checking and validating the user's email address. set a fake email for the `noauth` user to address that:

Screen Shot 2020-03-11 at 4 48 33 PM

**Comment by bruth** [comment]
Hm. I can't reproduce now.. maybe it was a strange cache issue for me. It is working great now!

**Comment by hessam61** [comment]
Perfect, I guess I'll keep the fake email address for the noauth user just to be on the safe side.

**Comment by murphyke** [comment]
@hessam61 Sorry, I got distracted reading a [lengthy analysis of COVID-19](https://medium.com/@tomaspueyo/coronavirus-act-today-or-people-will-die-f4d3d9cd99ca). I'll have to finish my review tomorrow.

**Comment by jglev** [comment]
> @bruth That is interesting, docker-compose takes a long time for me too. This wasn't the case few hours ago, it runs but takes a really long time now. I wonder if one of the NPM repositories has a slow connection.

> Hm. I can't reproduce now.. maybe it was a strange cache issue for me. It is working great now!

Just chiming in to say, could this be related to the issue that you and I noticed a month or so ago for the Labs project(?), @bruth, where Debian repo. package downloads were sometimes failing during Docker builds, but only when we were connected to the VPN? That is to say, what you both describe is for a different package management system, but strikes me as possibly being VPN-related, like that other instance. 🤷‍♀

**Comment by hessam61** [comment]
> > @bruth That is interesting, docker-compose takes a long time for me too. This wasn't the case few hours ago, it runs but takes a really long time now. I wonder if one of the NPM repositories has a slow connection.
> 
> > Hm. I can't reproduce now.. maybe it was a strange cache issue for me. It is working great now!
> 
> Just chiming in to say, could this be related to the issue that you and I noticed a month or so ago for the Labs project(?), @bruth, where Debian repo. package downloads were sometimes failing during Docker builds, but only when we were connected to the VPN? That is to say, what you both describe is for a different package management system, but strikes me as possibly being VPN-related, like that other instance. 🤷‍♀

I totally believe that as I can't reproduce it today. Having said that, very few times I ran into networking issues with some npm packages/repo in the past causing really long build time.

**Review by bruth — APPROVED** [review]
This looks good to me now that we can't reproduce the `userEmail` issue. One thing to consider is that this feature should get proposed to upstream. I am not sure if he would want to the Dockerfile.dev and docker-compose file since that doesn't follow his current dev, build, test process. However we could let him decide that.

**Comment by hessam61** [comment]
> This looks good to me now that we can't reproduce the `userEmail` issue. One thing to consider is that this feature should get proposed to upstream. I am not sure if he would want to the Dockerfile.dev and docker-compose file since that doesn't follow his current dev, build, test process. However we could let him decide that.

@bruth I agree. I was planning to remove that when we're ready to submit PR to upstream. It was just a way for easier system testing internally. In fact, I think I'll create a new branch from this and remove dockerfile.dev and associated docker-compose, and submit that as PR to upstream. I believe this is how @murphyke done it as well? Once that's done and approved by the maintainer, we rebase our master.

**Comment by bruth** [comment]
Sounds great!

**Review by murphyke — APPROVED** [review]
LGTM, nice!

**Comment by bruth** [comment]
@hessam61 Kevin approved so feel free to merge!

**Comment by hessam61** [comment]
Great, I'll submit PR to upstream next.

**Pull request merged**
